dashboard/app/models/course.rb

```diff
@@ -121,4 +121,59 @@ def summarize
       end
     }
   end
+
+  @@course_cache = nil
+  COURSE_CACHE_KEY = 'course-cache'.freeze
+
+  # TODO(bjvanminnen): share this somehow with script.rb?
+  def self.should_cache?
+    return false if Rails.application.config.levelbuilder_mode
+    return false if ENV['UNIT_TEST'] || ENV['CI']
+    true
+  end
+
+  # generates our course_cache from what is in the Rails cache
+  def self.course_cache_from_cache
+    Course.connection
+    # make sure possible loaded objects are completely loaded
+    [CourseScript, Plc::Course].each(&:new)
+    Rails.cache.read COURSE_CACHE_KEY
+  end
+
+  def self.course_cache_from_db
+    {}.tap do |cache|
+      Course.all.pluck(:id).each do |course_id|
+        course = get_without_cache(course_id)
+        cache[course.name] = course
+        cache[course.id.to_s] = course
+      end
+    end
+  end
+
+  def self.course_cache_to_cache
+    Rails.cache.write(COURSE_CACHE_KEY, course_cache_from_db)
+  end
+
+  def self.course_cache
+    return nil unless should_cache?
+    @@course_cache ||=
+      course_cache_from_cache || course_cache_from_db
+  end
+
+  def self.get_without_cache(id)
```

Comment:
nit: Though this is simply duplication of the existing paradigm, the method API and the method implementation would be better documented if `id` were replaced with `id_or_name` or `course_id_or_course_name`. Also below.